### PR TITLE
New i3 extractors for high energy, small bug fixes

### PR DESCRIPTION
**Pull request by RasmusOrsoe**
This PR contains 3 elements; 1 bug fix (duct tape), implementation of Kunals pulse flags and two new i3extractors specific to high energy. 

**Bug Fix**
The calculation of `track_energy` in the `I3TruthExtractor` is not general - it fails on high energy events with 
```
File "/lustre/hpc/hep/pcs557/github/graphnet/src/graphnet/data/extractors/i3truthextractor.py", line 352, in <genexpr>
    energy_track = sum(track.total_energy for track in tracks)
RuntimeError: "Hadrons" has no mass implemented. Cannot get total energy. (in double I3Particle::GetTotalEnergy() const)
```
As an intermediate solution, while a proper understanding of why track energy calculation has to be handled differently between the energy regimes, I've implemented the following try/except fix. See #282 
```
try:
                (
                    energy_track,
                    inelasticity,
                ) = self._get_primary_track_energy_and_inelasticity(frame)
            except RuntimeError: # track energy fails on northeren tracks with ""Hadrons" has no mass implemented. Cannot get total energy."
                energy_track = padding_value,
                inelasticity = padding_value,
```
 **implementation of Kunals pulse flags**
`I3FeatureExtractorIceCube86` now returns booleans "is_bright_dom", "is_bad_dom", "is_errata_dom", "is_saturated_dom"  along with the ordinary pulsemap features. It returns `-1` on these quantities if the respective dom lists are not available in the frame.

**Two new i3extractors specific to high energy.**
One extractor gets direction reconstructions from SplineMPEIC and the other gets rather use-case specific reconstructed quantities from a DNN-based energy reco algorithm from TUM.

### CODE REVIEW

File: src/graphnet/data/extractors/i3featureextractor.py [hunk 1]
```diff
@@ -87,7 +136,19 @@ def __call__(self, frame) -> dict:
...
+                )  # 0 or 1 or padding_value if list is not in file
+                output["is_bad_dom"].append(
+                    is_bad_dom
+                )  # 0 or 1 or padding_value if list is not in file
+                output["is_saturated_dom"].append(
+                    is_saturated_dom
+                )  # 0 or 1 or padding_value if list is not in file
+                output["is_errata_dom"].append(
+                    is_errata_dom
+                )  # 0 or 1 or padding_value if list is not in file
```
Comment (asogaard) [inline src/graphnet/data/extractors/i3featureextractor.py:151]:
```suggestion
                # DOM flags
                output["is_bright_dom"].append(is_bright_dom)
                output["is_bad_dom"].append(is_bad_dom)
                output["is_saturated_dom"].append(is_saturated_dom)
                output["is_errata_dom"].append(is_errata_dom)
```

File: src/graphnet/data/extractors/i3featureextractor.py [hunk 2]
```diff
@@ -76,6 +95,36 @@ def __call__(self, frame) -> dict:
...
+            else:
+                is_saturated_dom = -1
+
+            if calibration_errata:
+                if om_key in calibration_errata:
+                    is_errata_dom = 1
+                else:
+                    is_errata_dom = 0
+            else:
+                is_errata_dom = -1
```
Comment (asogaard) [inline src/graphnet/data/extractors/i3featureextractor.py:126]: Just trying to make these a little bit more compact.
```suggestion
            # DOM flags
            if bright_doms:
                is_bright_dom = 1 if om_key in bright_doms else 0
            else:
                is_bright_dom = -1

            if bad_doms:
                is_bad_dom = 1 if om_key in bad_doms else 0
            else:
                is_bad_dom = -1

            if saturation_windows:
                is_saturated_dom = 1 if om_key in saturation_windows else 0
            else:
                is_saturated_dom = -1

            if calibration_errata:
                is_errata_dom = 1 if om_key in calibration_errata else 0
            else:
                is_errata_dom = -1
```

File: src/graphnet/data/extractors/i3splinempeextractor.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+            output.update(
+                {
+                    "zenith_spline_mpe_ic": frame[
+                        "SplineMPEIC"
+                    ].dir.zenith,
+                    "azimuth_spline_mpe_ic": frame[
+                        "SplineMPEIC"
+                    ].dir.azimuth,
+                }
+            )
```
Comment (asogaard) [inline src/graphnet/data/extractors/i3splinempeextractor.py:20]: Man, sometimes `black` gets on my nerves...

File: src/graphnet/data/extractors/i3splinempeextractor.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+from graphnet.data.extractors.i3extractor import I3Extractor
+
+class I3SplineMPEICExtractor(I3Extractor):
+    def __init__(self, name="spline_mpe_ic"):
+        super().__init__(name)
+
+    def __call__(self, frame) -> dict:
+        """Extracts TUM DNN Recos and associated variables"""
```
Comment (asogaard) [inline src/graphnet/data/extractors/i3splinempeextractor.py:8]:
```suggestion
        """Extracts SplineMPE pointing predictions."""
```

**Review by asogaard — CHANGES_REQUESTED** [review]
Hi @RasmusOrsoe,

Thanks for these fixes! I have added a few suggestions that are mostly cosmetic. Let me know what you think, and then I'm sure we can merge this in no time. :)

File: src/graphnet/data/extractors/i3splinempeextractor.py (hunk 3 above)
Comment (RasmusOrsoe) [reply src/graphnet/data/extractors/i3splinempeextractor.py:20]: I feel you

File: src/graphnet/data/extractors/i3truthextractor.py [hunk 5]
```diff
@@ -146,10 +146,15 @@ def __call__(self, frame, padding_value=-1) -> dict:
...
-            ) = self._get_primary_track_energy_and_inelasticity(frame)
+            try:
+                (
+                    energy_track,
+                    inelasticity,
+                ) = self._get_primary_track_energy_and_inelasticity(frame)
+            except RuntimeError: # track energy fails on northeren tracks with ""Hadrons" has no mass implemented. Cannot get total energy."
+                energy_track = padding_value,
+                inelasticity = padding_value,
+
```
[the remark this answers is not in the document]
Comment (RasmusOrsoe) [reply src/graphnet/data/extractors/i3truthextractor.py:157]: I'm not entirely sure about this. I'd rather not change anything in the computation of the truth variable as I don't fully understand it. I don't know if your suggestion produces the correct answer. Does it?

**Comment by RasmusOrsoe** [comment]
@asogaard I think everything here is in line, except the issue of track energy. Can you comment?

**Review by asogaard — APPROVED** [review]
Hi @RasmusOrsoe,
I think this looks fine! :)

**Pull request merged**